## Practica 1

In [2]:
import pandas as pd

In [3]:
spot = pd.read_csv(r"C:\Users\Usuario\Downloads\spotify.csv")
zips = pd.read_csv(r"C:\Users\Usuario\Downloads\zips.csv")

C:\Users\Usuario\AppData\Local\Temp\ipykernel_11560\3701591321.py:1: DtypeWarning: Columns (18,19) have mixed types. Specify dtype option on import or set low_memory=False.
  spot = pd.read_csv(r"C:\Users\Usuario\Downloads\spotify.csv")


## Mostramos los datos

In [4]:
spot

acousticness                                            artists  \
0              0.764                               ['Shel Silverstein']   
1              0.917                                   ['Mose Allison']   
2              0.926                                            ['水柳仙']   
3              0.976                                    ['Dean Martin']   
4              0.892                                      ['Doris Day']   
...              ...                                                ...   
170344         0.645                                    ['The Clovers']   
170345         0.671                                   ['Ricky Nelson']   
170346         0.740  ['The Stanley Brothers', 'The Clinch Mountain ...   
170347         0.486                                     ['Art Pepper']   
170348         0.989  ['James McGranahan', 'Eugene Ormandy', 'Philad...   

        danceability  duration_ms  energy  explicit                      id  \
0              0.669       210387  0.1720       NaN  1r3N7MoEGPHJyCjBlWCsx3   
1              0.693       149547  0.2410       NaN  2RL8EkOOu0QhAikhUUvIRi   
2              0.183       207173  0.0854       0.0  3kKKEghp2EMh8992pWFylr   
3              0.472       182653  0.0314       NaN  0j3ideT0tRuIyRhOOUQjYa   
4              0.243       171400  0.3360       NaN  0od9zbD6Bcc8b7dCMRFWBZ   
...              ...          ...     ...       ...                     ...   
170344         0.519       153533  0.0825       0.0  2JUy9z8u8rxddAEGCAzpec   
170345         0.700       132200  0.7380       0.0  2OBi2t8GK7X0GQdu4v5y91   
170346         0.505       159507  0.2860       0.0  2RHOe9Z6aSVptkk76VuTgQ   
170347         0.711       316227  0.2550       0.0  2TfzIScmcxkOJNsrLhMl11   
170348         0.194       365627  0.1600       0.0  2W6D4C326XM8XEsiewAXjK   

        instrumentalness   key  liveness  loudness  mode  \
0               0.000000   2.0    0.1280   -13.075     1   
1               0.000014   NaN    0.0824   -14.605     0   
2               0.704000   7.0    0.1120   -26.598     1   
3               0.000015   7.0    0.1500   -18.006     1   
4               0.000049   2.0    0.1310    -9.812     0   
...                  ...   ...       ...       ...   ...   
170344          0.000000   8.0    0.1010   -14.650     1   
170345          0.000004   8.0    0.2210   -11.116     1   
170346          0.000000  10.0    0.3330   -11.885     1   
170347          0.043600   NaN    0.0871   -11.397     1   
170348          0.358000   8.0    0.1780   -22.398     1   

                                       name  popularity         release_date  \
0                               The Unicorn        12.0  1962-10-13 00:00:00   
1                               If You Live        19.0  1963-04-14 00:00:00   
2                                       鴛鴦夢        23.0  1963-11-14 00:00:00   
3                             Fools Rush In        21.0  1964-06-07 00:00:00   
4                       The Christmas Waltz        22.0  1964-08-07 00:00:00   
...                                     ...         ...                  ...   
170344                       Fool Fool Fool        14.0  1959-10-15 00:00:00   
170345  Just A Little Too Much - Remastered        20.0  1959-04-16 00:00:00   
170346              This Wicked Path Of Sin        11.0  1959-01-25 00:00:00   
170347              Walkin' - Original Take        17.0  1959-06-25 00:00:00   
170348                 O, My Father - Voice        12.0  1959-09-05 00:00:00   

        speechiness    tempo  valence zip Code genero  
0            0.3270   97.079    0.333  38ee029    NaN  
1            0.0350  120.890    0.715   9p2304    NaN  
2            0.0392  101.482    0.131  2166ee6    NaN  
3            0.0364  135.555    0.194    67t06    NaN  
4            0.0282   89.158    0.238   804p53    NaN  
...             ...      ...      ...      ...    ...  
170344       0.0380   78.755    0.754   2t7616   20.0  
170345       0.0350   91.659

In [5]:
zips

zip       lat        lng            city state_id      state_name  \
0        601  18.18004  -66.75218      Adjunta28s       PR  Puerto 345Rico   
1        602  18.36073  -67.17517       A543guada       PR  P733uerto Rico   
2        603  18.45439  -67.12202    Ag188uadilla       PR  Pue888rto Rico   
3        606  18.16724  -66.93828      Ma297ricao       PR  Puerto Ric683o   
4        610  18.29032  -67.12243       Anasc449o       PR  P753uerto Rico   
...      ...       ...        ...             ...      ...             ...   
33094  99923  55.97796 -130.03671        H481yder       AK       A540laska   
33095  99925  55.55767 -132.97627      442Klawock       AK       Alask533a   
33096  99926  55.12617 -131.48928   M165etlakatla       AK       797Alaska   
33097  99927  56.25100 -133.37571  Po905int Baker       AK       Alask349a   
33098  99929  56.36950 -131.93648     Wrange821ll       AK       Alas486ka   

       zcta  parent_zcta  population  density  county_fips  \
0      True          NaN       17242    111.4        72001   
1      True          NaN       38442    523.5        72003   
2      True          NaN       48814    667.9        72005   
3      True          NaN        6437     60.4        72093   
4      True          NaN       27073    312.0        72011   
...     ...          ...         ...      ...          ...   
33094  True          NaN          15      2.1         2198   
33095  True          NaN         927      5.7         2198   
33096  True          NaN        1635      4.2         2198   
33097  True          NaN          38      0.2         2198   
33098  True          NaN        2484      0.4         2275   

                 county_name                             county_weights  \
0                   Adjuntas               {'72001':99.43,'72141':0.57}   
1                     Aguada                              {'72003':100}   
2                  Aguadilla                              {'72005':100}   
3                    Maricao  {'72093':94.88,'72121':1.35,'72153':3.78}   
4                     Añasco               {'72003':0.55,'72011':99.45}   
...                      ...                                        ...   
33094  Prince of Wales-Hyder                              {'02198':100}   
33095  Prince of Wales-Hyder                              {'02198':100}   
33096  Prince of Wales-Hyder                              {'02198':100}   
33097  Prince of Wales-Hyder                              {'02198':100}   
33098               Wrangell                              {'02275':100}   

                  county_names_all    county_fips_all  imprecise  military  \
0                  Adjuntas|Utuado        72001|72141      False     False   
1                           Aguada              72003      False     False   
2                        Aguadilla              72005      False     False   
3      Maricao|Yauco|Sabana Grande  72093|72153|72121      False     False   
4                    Añasco|Aguada        72011|72003      False     False   
...                            ...                ...        ...       ...   
33094        Prince of Wales-Hyder              02198      False     False   
33095        Prince of Wales-Hyder              02198      False     False   
33096        Prince of Wales-Hyder              02198      False     False   
33097        Prince of Wales-Hyder              02198      False     False   
33098                     Wrangell              02275      False     False   

                  timezone  c_lng  
0      America/Puerto_Rico    NaN  
1      America/Puerto_Rico    NaN  
2      America/Puerto_Rico    NaN  
3      America/Puerto_Rico    NaN  
4      America/Puerto_Rico    NaN  
...                    ...    ...  
33094        America/Sitka    NaN  
33095        America/Sitka    NaN  
33096   America/Metlakatla    NaN  
33097        America/Sitka    NaN  
33098        America/Sitka    NaN  

[33099 rows x 19 columns]

In [6]:
spot.head()


acousticness               artists  danceability  duration_ms  energy  \
0         0.764  ['Shel Silverstein']         0.669       210387  0.1720   
1         0.917      ['Mose Allison']         0.693       149547  0.2410   
2         0.926               ['水柳仙']         0.183       207173  0.0854   
3         0.976       ['Dean Martin']         0.472       182653  0.0314   
4         0.892         ['Doris Day']         0.243       171400  0.3360   

   explicit                      id  instrumentalness  key  liveness  \
0       NaN  1r3N7MoEGPHJyCjBlWCsx3          0.000000  2.0    0.1280   
1       NaN  2RL8EkOOu0QhAikhUUvIRi          0.000014  NaN    0.0824   
2       0.0  3kKKEghp2EMh8992pWFylr          0.704000  7.0    0.1120   
3       NaN  0j3ideT0tRuIyRhOOUQjYa          0.000015  7.0    0.1500   
4       NaN  0od9zbD6Bcc8b7dCMRFWBZ          0.000049  2.0    0.1310   

   loudness  mode                 name  popularity         release_date  \
0   -13.075     1          The Unicorn        12.0  1962-10-13 00:00:00   
1   -14.605     0          If You Live        19.0  1963-04-14 00:00:00   
2   -26.598     1                  鴛鴦夢        23.0  1963-11-14 00:00:00   
3   -18.006     1        Fools Rush In        21.0  1964-06-07 00:00:00   
4    -9.812     0  The Christmas Waltz        22.0  1964-08-07 00:00:00   

   speechiness    tempo  valence zip Code genero  
0       0.3270   97.079    0.333  38ee029    NaN  
1       0.0350  120.890    0.715   9p2304    NaN  
2       0.0392  101.482    0.131  2166ee6    NaN  
3       0.0364  135.555    0.194    67t06    NaN  
4       0.0282   89.158    0.238   804p53    NaN

In [7]:
zips.head()

zip       lat       lng          city state_id      state_name  zcta  \
0  601  18.18004 -66.75218    Adjunta28s       PR  Puerto 345Rico  True   
1  602  18.36073 -67.17517     A543guada       PR  P733uerto Rico  True   
2  603  18.45439 -67.12202  Ag188uadilla       PR  Pue888rto Rico  True   
3  606  18.16724 -66.93828    Ma297ricao       PR  Puerto Ric683o  True   
4  610  18.29032 -67.12243     Anasc449o       PR  P753uerto Rico  True   

   parent_zcta  population  density  county_fips county_name  \
0          NaN       17242    111.4        72001    Adjuntas   
1          NaN       38442    523.5        72003      Aguada   
2          NaN       48814    667.9        72005   Aguadilla   
3          NaN        6437     60.4        72093     Maricao   
4          NaN       27073    312.0        72011      Añasco   

                              county_weights             county_names_all  \
0               {'72001':99.43,'72141':0.57}              Adjuntas|Utuado   
1                              {'72003':100}                       Aguada   
2                              {'72005':100}                    Aguadilla   
3  {'72093':94.88,'72121':1.35,'72153':3.78}  Maricao|Yauco|Sabana Grande   
4               {'72003':0.55,'72011':99.45}                Añasco|Aguada   

     county_fips_all  imprecise  military             timezone  c_lng  
0        72001|72141      False     False  America/Puerto_Rico    NaN  
1              72003      False     False  America/Puerto_Rico    NaN  
2              72005      False     False  America/Puerto_Rico    NaN  
3  72093|72153|72121      False     False  America/Puerto_Rico    NaN  
4        72011|72003      False     False  America/Puerto_Rico    NaN

In [8]:
spot.dtypes

acousticness        float64
artists              object
danceability        float64
duration_ms           int64
energy              float64
explicit            float64
id                   object
instrumentalness    float64
key                 float64
liveness            float64
loudness            float64
mode                  int64
name                 object
popularity          float64
release_date         object
speechiness         float64
tempo               float64
valence             float64
zip Code             object
genero               object
dtype: object

### a) Realizamos el etiquetado de las variables

In [18]:
c_feats = ["acousticnes","danceability","duration_ms", "energy", "instrumentalness", "liveness", "loudness", "speechiness", "tempo", "valence", "zip Code"]
d_feats = ["explicit", "key", "mode", "popularity", "genero"]
t_feats = ["release_date"]
s_feats = ["artists", "id", "name"]

c_feats_new = ["c_" + x for x in c_feats]
d_feats_new = ["d_" + x for x in d_feats]
t_feats_new = ["t_" + x for x in t_feats]
s_feats_new = ["s_" + x for x in s_feats]

In [19]:
spot.rename(columns = dict(zip(c_feats, c_feats_new)), inplace = True)
spot.rename(columns = dict(zip(d_feats, d_feats_new)), inplace = True)
spot.rename(columns = dict(zip(t_feats, t_feats_new)), inplace = True)
spot.rename(columns = dict(zip(s_feats, s_feats_new)), inplace = True)

In [20]:
spot.head() #Mostramos las columnas en el df

acousticness             s_artists  c_danceability  c_duration_ms  \
0         0.764  ['Shel Silverstein']           0.669         210387   
1         0.917      ['Mose Allison']           0.693         149547   
2         0.926               ['水柳仙']           0.183         207173   
3         0.976       ['Dean Martin']           0.472         182653   
4         0.892         ['Doris Day']           0.243         171400   

   c_energy  d_explicit                    s_id  c_instrumentalness  d_key  \
0    0.1720         NaN  1r3N7MoEGPHJyCjBlWCsx3            0.000000    2.0   
1    0.2410         NaN  2RL8EkOOu0QhAikhUUvIRi            0.000014    NaN   
2    0.0854         0.0  3kKKEghp2EMh8992pWFylr            0.704000    7.0   
3    0.0314         NaN  0j3ideT0tRuIyRhOOUQjYa            0.000015    7.0   
4    0.3360         NaN  0od9zbD6Bcc8b7dCMRFWBZ            0.000049    2.0   

   c_liveness  c_loudness  d_mode               s_name  d_popularity  \
0      0.1280     -13.075       1          The Unicorn          12.0   
1      0.0824     -14.605       0          If You Live          19.0   
2      0.1120     -26.598       1                  鴛鴦夢          23.0   
3      0.1500     -18.006       1        Fools Rush In          21.0   
4      0.1310      -9.812       0  The Christmas Waltz          22.0   

        t_release_date  c_speechiness  c_tempo  c_valence s_zip Code s_genero  
0  1962-10-13 00:00:00         0.3270   97.079      0.333    38ee029      NaN  
1  1963-04-14 00:00:00         0.0350  120.890      0.715     9p2304      NaN  
2  1963-11-14 00:00:00         0.0392  101.482      0.131    2166ee6      NaN  
3  1964-06-07 00:00:00         0.0364  135.555      0.194      67t06      NaN  
4  1964-08-07 00:00:00         0.0282   89.158      0.238     804p53      NaN

In [12]:
spot.columns #Mostramos las etiquetas

Index(['acousticness', 's_artists', 'c_danceability', 'c_duration_ms',
       'c_energy', 'd_explicit', 's_id', 'c_instrumentalness', 'd_key',
       'c_liveness', 'c_loudness', 'd_mode', 's_name', 'd_popularity',
       't_release_date', 'c_speechiness', 'c_tempo', 'c_valence', 's_zip Code',
       's_genero'],
      dtype='object')

## Imprimimos y eliminamos los duplicados

In [14]:
spot.duplicated().sum()

np.int64(364)

In [15]:
spot.drop_duplicates(inplace=True)

In [16]:
spot.duplicated().sum() #Nos aseguramos de que se hayan eliminado

np.int64(0)

In [15]:
spot.reset_index(drop=True,inplace=True) #Reseteamos index para que todo se acomode

In [16]:
spot = spot.reset_index(drop=True)

In [17]:
spot.shape

(169985, 20)

In [18]:
spot.dtypes

acousticness          float64
s_artists              object
c_danceability        float64
c_duration_ms           int64
c_energy              float64
d_explicit            float64
s_id                   object
c_instrumentalness    float64
d_key                 float64
c_liveness            float64
c_loudness            float64
d_mode                  int64
s_name                 object
d_popularity          float64
t_release_date         object
c_speechiness         float64
c_tempo               float64
c_valence             float64
s_zip Code             object
s_genero               object
dtype: object

In [19]:
import unicodedata
import re

In [20]:
def clean_text(text, pattern="[^a-zA-Z0-9 ]"):
    
    cleaned_text = unicodedata.normalize('NFD', text).encode('ascii', 'ignore')
    cleaned_text = re.sub(pattern, " ", cleaned_text.decode("utf-8"), flags=re.UNICODE)
    cleaned_text = u' '.join(cleaned_text.lower().strip().split())
    return cleaned_text

#Creamos esta función que nos sirve para estandarizar las variables de texto

In [21]:
spot["s_artists"] = spot["s_artists"].map(lambda x:clean_text(x.lower()).replace("\n",""))

In [22]:

spot["s_id"] = spot["s_id"].map(lambda x:clean_text(x.lower()).replace("\n",""))

In [23]:
spot["s_name"] = spot["s_name"].map(lambda x:clean_text(x.lower()).replace("\n",""))

In [25]:

spot["d_genero"] = spot["d_genero"].astype(str).map(lambda x: clean_text(x.lower()).replace("\n", ""))

In [26]:
spot = spot.reset_index(drop=True)

In [27]:
spot.shape

(169985, 20)

## (c) Realizamos la completitud de las variables

In [28]:
comple = pd.DataFrame(spot.isnull().sum()) #Creamos un dataframe con pd.DataFrame con el n de nulos

In [29]:
comple.reset_index(inplace=True)

In [30]:
comple = comple.rename(columns={"index":"columna",0:"total"}) #Renombramos columnas

In [31]:
comple["completitud"]=(1-comple["total"]/spot.shape[0])*100 #Calculamos %

In [32]:
spot.shape

(169985, 20)

In [33]:
comple=comple.sort_values(by="completitud",ascending=True) #Ordenamos los valores

In [34]:
comple

columna  total  completitud
5           d_explicit  62905    62.993794
8                d_key  35759    78.963438
13        d_popularity   3501    97.940407
0         acousticness      0   100.000000
3        c_duration_ms      0   100.000000
4             c_energy      0   100.000000
2       c_danceability      0   100.000000
1            s_artists      0   100.000000
7   c_instrumentalness      0   100.000000
6                 s_id      0   100.000000
10          c_loudness      0   100.000000
9           c_liveness      0   100.000000
11              d_mode      0   100.000000
12              s_name      0   100.000000
14      t_release_date      0   100.000000
15       c_speechiness      0   100.000000
16             c_tempo      0   100.000000
17           c_valence      0   100.000000
18          s_zip Code      0   100.000000
19            s_genero      0   100.000000

## (d) Eliminamos las variables que tengan el 20% o más de valores ausentes

In [35]:
spot.drop(columns = ["d_explicit","d_key"], inplace = True)

In [36]:
spot.shape

(169985, 18)

In [37]:

spot.head()

acousticness         s_artists  c_danceability  c_duration_ms  c_energy  \
0         0.764  shel silverstein           0.669         210387    0.1720   
1         0.917      mose allison           0.693         149547    0.2410   
2         0.926                             0.183         207173    0.0854   
3         0.976       dean martin           0.472         182653    0.0314   
4         0.892         doris day           0.243         171400    0.3360   

                     s_id  c_instrumentalness  c_liveness  c_loudness  d_mode  \
0  1r3n7moegphjycjblwcsx3            0.000000      0.1280     -13.075       1   
1  2rl8ekoou0qhaikhuuviri            0.000014      0.0824     -14.605       0   
2  3kkkeghp2emh8992pwfylr            0.704000      0.1120     -26.598       1   
3  0j3idet0truiyrhoouqjya            0.000015      0.1500     -18.006       1   
4  0od9zbd6bcc8b7dcmrfwbz            0.000049      0.1310      -9.812       0   

                s_name  d_popularity       t_release_date  c_speechiness  \
0          the unicorn          12.0  1962-10-13 00:00:00         0.3270   
1          if you live          19.0  1963-04-14 00:00:00         0.0350   
2                               23.0  1963-11-14 00:00:00         0.0392   
3        fools rush in          21.0  1964-06-07 00:00:00         0.0364   
4  the christmas waltz          22.0  1964-08-07 00:00:00         0.0282   

   c_tempo  c_valence s_zip Code s_genero  
0   97.079      0.333    38ee029      nan  
1  120.890      0.715     9p2304      nan  
2  101.482      0.131    2166ee6      nan  
3  135.555      0.194      67t06      nan  
4   89.158      0.238     804p53      nan

In [38]:
for col in spot.filter(like = "s_").columns:
    print(col)
    display(spot[col].value_counts())
    print("\n")

s_artists


s_artists
                                                                                                                             3122
francisco canaro                                                                                                              947
ignacio corsini                                                                                                               626
frank sinatra                                                                                                                 592
bob dylan                                                                                                                     539
                                                                                                                             ... 
engelbert humperdinck herbert von karajan josef metternich philharmonia orchestra                                               1
richard wagner solisten des bayerischen rundfunkchors bayerischer rundfunkchor b



s_id


s_id
6tmvnrsyktvrfhuupnn7jk    3
4cnphg2qwjju5gwwbo1ocx    3
72f1vdd6velfzybk1wkggy    3
72vd6lhg8kohmxjgqkkzbx    3
6ojjveoywjdit76y0pxpxw    3
                         ..
2hcdmlsa6rxw1yfbbt45q6    1
4bloigb3nt7diwhgikd9c8    1
0lu2tz2hnhymqzyazqnzes    1
0piqpvfz3fswmtseerwiwq    1
49ueuxlw7gxacgeeg4ewnp    1
Name: count, Length: 169909, dtype: int64



s_name


s_name
                                         1826
summertime                                 63
overture                                   43
home                                       41
stay                                       36
                                         ... 
airegin                                     1
gypsy some people                           1
the moon her majesty with steve allen       1
jamaica farewell live                       1
high school confidential 1964 version       1
Name: count, Length: 126205, dtype: int64



s_zip Code


s_zip Code
30533      16
21122      16
68106      16
39348      16
85286      16
           ..
8te6321     1
232u34      1
6te3362     1
4te3822     1
7875p4      1
Name: count, Length: 37217, dtype: int64



s_genero


s_genero
47 0                      33872
20 0                      31088
50 0                      25038
12 0                      22655
16 0                      22470
21 0                      22286
30 0                       3308
49 0                       3264
48 0                       3217
nan                        1698
3 0                         546
2 0                         361
1 0                         171
3 sure                        1
ianca                         1
2 3 3 1                       1
fantasiestucke                1
2woitschach                   1
cristian                      1
bebamos juntos                1
polonaise                     1
6n6tifz9vltsoixkj8qkrd        1
6l63vw0pibdm1hdsboqnom        1
                              1
Name: count, dtype: int64

In [39]:
spot['s_zip Code']


0         38ee029
1          9p2304
2         2166ee6
3           67t06
4          804p53
           ...   
169980     2t7616
169981     4p2071
169982     7t6827
169983     9u2233
169984     214te4
Name: s_zip Code, Length: 169985, dtype: object

In [40]:
spot['s_zip Code'].value_counts()

s_zip Code
30533      16
21122      16
68106      16
39348      16
85286      16
           ..
8te6321     1
232u34      1
6te3362     1
4te3822     1
7875p4      1
Name: count, Length: 37217, dtype: int64

In [41]:
spot['s_zip Code'] = spot['s_zip Code'].astype('str')

In [42]:
spot['s_zip Code'].value_counts()

s_zip Code
30533      16
21122      16
68106      16
39348      16
85286      16
           ..
8te6321     1
232u34      1
6te3362     1
4te3822     1
7875p4      1
Name: count, Length: 37217, dtype: int64

filtra el DataFrame spot para seleccionar todas las filas cuyo valor en la columna 's_zip Code' no está compuesto únicamente por dígitos.

In [43]:
spot[~spot['s_zip Code'].map(lambda x: x.isdigit())]

acousticness                                          s_artists  \
0              0.764                                   shel silverstein   
1              0.917                                       mose allison   
2              0.926                                                      
3              0.976                                        dean martin   
4              0.892                                          doris day   
...              ...                                                ...   
169980         0.645                                        the clovers   
169981         0.671                                       ricky nelson   
169982         0.740      the stanley brothers the clinch mountain boys   
169983         0.486                                         art pepper   
169984         0.989  james mcgranahan eugene ormandy philadelphia o...   

        c_danceability  c_duration_ms  c_energy                    s_id  \
0                0.669         210387    0.1720  1r3n7moegphjycjblwcsx3   
1                0.693         149547    0.2410  2rl8ekoou0qhaikhuuviri   
2                0.183         207173    0.0854  3kkkeghp2emh8992pwfylr   
3                0.472         182653    0.0314  0j3idet0truiyrhoouqjya   
4                0.243         171400    0.3360  0od9zbd6bcc8b7dcmrfwbz   
...                ...            ...       ...                     ...   
169980           0.519         153533    0.0825  2juy9z8u8rxddaegcazpec   
169981           0.700         132200    0.7380  2obi2t8gk7x0gqdu4v5y91   
169982           0.505         159507    0.2860  2rhoe9z6asvptkk76vutgq   
169983           0.711         316227    0.2550  2tfziscmcxkojnsrlhml11   
169984           0.194         365627    0.1600  2w6d4c326xm8xesiewaxjk   

        c_instrumentalness  c_liveness  c_loudness  d_mode  \
0                 0.000000      0.1280     -13.075       1   
1                 0.000014      0.0824     -14.605       0   
2                 0.704000      0.1120     -26.598       1   
3                 0.000015      0.1500     -18.006       1   
4                 0.000049      0.1310      -9.812       0   
...                    ...         ...         ...     ...   
169980            0.000000      0.1010     -14.650       1   
169981            0.000004      0.2210     -11.116       1   
169982            0.000000      0.3330     -11.885       1   
169983            0.043600      0.0871     -11.397       1   
169984            0.358000      0.1780     -22.398       1   

                                   s_name  d_popularity       t_release_date  \
0                             the unicorn          12.0  1962-10-13 00:00:00   
1                             if you live          19.0  1963-04-14 00:00:00   
2                                                  23.0  1963-11-14 00:00:00   
3                           fools rush in          21.0  1964-06-07 00:00:00   
4                     the christmas waltz          22.0  1964-08-07 00:00:00   
...                                   ...           ...                  ...   
169980                     fool fool fool          14.0  1959-10-15 00:00:00   
169981  just a little too much remastered          20.0  1959-04-16 00:00:00   
169982            this wicked path of sin          11.0  1959-01-25 00:00:00   
169983               walkin original take          17.0  1959-06-25 00:00:00   
169984                  o my father voice          12.0  1959-09-05 00:00:00   

        c_speechiness  c_tempo  c_valence s_zip Code s_genero  
0              0.3270   97.079      0.333    38ee029      nan  
1              0.0350  120.890      0.715     9p2304      nan  
2              0.0392  101.482      0.131    2166ee6      nan  
3              0.0364  135.555      0.194      67t06      nan  
4              0.0282   89.158      0.238     804p53      nan  
...               ...      ...        ...        ...      ...  
169980         0.0380   78.755      0.754     2t7616

In [44]:
spot = spot[spot['s_zip Code'].map(lambda x: x.isdigit())]

In [45]:
spot.shape

(165644, 18)

In [46]:
spot['s_zip Code'].value_counts()

s_zip Code
39348    16
85286    16
21122    16
30533    16
68106    16
         ..
49233     1
92385     1
95237     1
58494     1
38011     1
Name: count, Length: 32886, dtype: int64

In [47]:
spot['s_zip Code']

24        13126
25        14223
26        84620
27        22824
28        27519
          ...  
168243    45830
168244    55814
168245    77564
168246    15082
168247    57472
Name: s_zip Code, Length: 165644, dtype: object

In [48]:
spot['s_zip Code'].dtypes

dtype('O')

In [49]:
spot


acousticness                                        s_artists  \
24          0.000331                                           slayer   
25          0.329000                                miguel mateos zas   
26          0.004680                                      paula abdul   
27          0.742000                                       henry fiol   
28          0.742000                                     eric clapton   
...              ...                                              ...   
168243      0.491000                                                    
168244      0.210000                                                    
168245      0.282000                                                    
168246      0.958000              franz joseph haydn pro arte quartet   
168247      0.948000  samuel barber leonard slatkin st louis symphony   

        c_danceability  c_duration_ms  c_energy                    s_id  \
24               0.296         400093     0.918  0dt3xql7ljtnyiztxy00yd   
25               0.548         290640     0.968  096vjnczzrwi4lluoggd9y   
26               0.731         231921     0.779  7xhyqboemdzxwuxpjf9h30   
27               0.708         305058     0.629  6m5z2pca6oun4l5n5kid3e   
28               0.598         216800     0.219  5tdewfbgnx7a7zd78tuwlz   
...                ...            ...       ...                     ...   
168243           0.729         101745     0.136  4q6g3smirzm5udxgew6jyx   
168244           0.628         113100     0.108  4qcmlwr23qjj8v0wejodeb   
168245           0.686         108000     0.176  4qgdvmu2fbayxoxryli4rm   
168246           0.569         208600     0.058  4qbv52okldyo4lsmljdxyg   
168247           0.185         294213     0.151  4qr6xmys4rtqt0nvuqqosq   

        c_instrumentalness  c_liveness  c_loudness  d_mode  \
24                0.178000      0.2530      -8.360       1   
25                0.172000      0.0759      -7.038       0   
26                0.001340      0.0525      -9.126       0   
27                0.000003      0.0858     -10.358       0   
28                0.056900      0.8790     -15.048       0   
...                    ...         ...         ...     ...   
168243            0.000000      0.1180     -16.814       0   
168244            0.000000      0.2140     -15.938       1   
168245            0.000000      0.1360     -16.794       0   
168246            0.000935      0.1650     -17.211       1   
168247            0.906000      0.1060     -16.431       0   

                                                   s_name  d_popularity  \
24                                     crypts of eternity          33.0   
25                                  mi sombra en la pared          45.0   
26                                            knocked out          32.0   
27                                                zumbale          46.0   
28      walkin blues acoustic live at mtv unplugged br...          45.0   
...                                                   ...           ...   
168243                                            8 4 9 1           0.0   
168244                                               40 3           0.0   
168245                                               47 2           0.0   
168246  haydn string quartet no 52 in e flat major op ...           0.0   
168247         symphony no 1 op 9 iv con moto passacaglia           0.0   

             t_release_date  c_speechiness  c_tempo  c_valence s_zip Code  \
24      1985-05-26 00:00:00         0.0693  101.916     0.1370      13126   
25      1986-10-02 00:00:00         0.0778  174.593     0.7860      14223   
26      1988-08-15 00:00:00         0.0454  116.233     0.9540      84620   
27      1991-04-16 00:00:00         0.0351   95.451     0.9150      22824   
28      1992-08-15 00:00:00         0.0476   85.330     0.4150      27519   
...                     ...            ...      ...        ...        ...   
168243  1935-01-25 00:00:00         0.9360   

In [50]:
spot['s_genero']

24         nan
25         nan
26         nan
27         nan
28         nan
          ... 
168243    12 0
168244    16 0
168245    20 0
168246    20 0
168247    12 0
Name: s_genero, Length: 165644, dtype: object

In [51]:
spot['s_genero'].value_counts()

s_genero
47 0                      32980
20 0                      30270
50 0                      24401
12 0                      22103
16 0                      21910
21 0                      21733
30 0                       3232
49 0                       3193
48 0                       3117
nan                        1642
3 0                         532
2 0                         351
1 0                         170
3 sure                        1
2 3 3 1                       1
fantasiestucke                1
2woitschach                   1
cristian                      1
bebamos juntos                1
polonaise                     1
6n6tifz9vltsoixkj8qkrd        1
6l63vw0pibdm1hdsboqnom        1
                              1
Name: count, dtype: int64

8 registros contienen letras, 1 solo contiene 1 espacio y 1642 son nan

In [52]:
# Mantiene las filas donde ningún carácter de 's_genero' es una letra
spot = spot[spot['s_genero'].map(lambda x: not any(c.isalpha() for c in str(x)))]

In [53]:
spot.shape

(163994, 18)

In [54]:
spot['s_name']

31                                          bab s uvula who
1680       piano sonata no 2 op 35 i grave doppio movimento
3921                                       sunless saturday
3922                                           read my lips
3923                                            rain forest
                                ...                        
168243                                              8 4 9 1
168244                                                 40 3
168245                                                 47 2
168246    haydn string quartet no 52 in e flat major op ...
168247           symphony no 1 op 9 iv con moto passacaglia
Name: s_name, Length: 163994, dtype: object

In [55]:
spot['s_name'] = spot['s_name'].str.lower().str.replace(r'[^a-zA-Z0-9\s]', '', regex=True) #Convertimos a minusculas con .str.lower()

C:\Users\Usuario\AppData\Local\Temp\ipykernel_5920\3006120894.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  spot['s_name'] = spot['s_name'].str.lower().str.replace(r'[^a-zA-Z0-9\s]', '', regex=True)


In [56]:
spot.shape

(163994, 18)

In [57]:
spot['s_name']

31                                          bab s uvula who
1680       piano sonata no 2 op 35 i grave doppio movimento
3921                                       sunless saturday
3922                                           read my lips
3923                                            rain forest
                                ...                        
168243                                              8 4 9 1
168244                                                 40 3
168245                                                 47 2
168246    haydn string quartet no 52 in e flat major op ...
168247           symphony no 1 op 9 iv con moto passacaglia
Name: s_name, Length: 163994, dtype: object

In [58]:
spot

acousticness                                        s_artists  \
31          0.000186                                        green day   
1680        0.990000                frederic chopin vladimir horowitz   
3921        0.000069                                         fishbone   
3922        0.068200                                   michael franks   
3923        0.936000                            paul speer david lanz   
...              ...                                              ...   
168243      0.491000                                                    
168244      0.210000                                                    
168245      0.282000                                                    
168246      0.958000              franz joseph haydn pro arte quartet   
168247      0.948000  samuel barber leonard slatkin st louis symphony   

        c_danceability  c_duration_ms  c_energy                    s_id  \
31               0.494         128200     0.932  2uarrcx71zl7jspun2kupz   
1680             0.272         473227     0.281  6wt8pknbsqnfqdoxgexvku   
3921             0.451         258440     0.970  4dtbf3w1v4nlzko6epwvue   
3922             0.861         222107     0.726  68lfky4tmwgkqfotnnez1j   
3923             0.605         258000     0.229  6fzfxmevhsdqkd8wjfp83n   
...                ...            ...       ...                     ...   
168243           0.729         101745     0.136  4q6g3smirzm5udxgew6jyx   
168244           0.628         113100     0.108  4qcmlwr23qjj8v0wejodeb   
168245           0.686         108000     0.176  4qgdvmu2fbayxoxryli4rm   
168246           0.569         208600     0.058  4qbv52okldyo4lsmljdxyg   
168247           0.185         294213     0.151  4qr6xmys4rtqt0nvuqqosq   

        c_instrumentalness  c_liveness  c_loudness  d_mode  \
31                0.000040      0.0248      -4.606       0   
1680              0.867000      0.0870     -16.228       0   
3921              0.103000      0.1360      -4.622       1   
3922              0.000130      0.0922     -12.759       1   
3923              0.860000      0.1000     -16.300       1   
...                    ...         ...         ...     ...   
168243            0.000000      0.1180     -16.814       0   
168244            0.000000      0.2140     -15.938       1   
168245            0.000000      0.1360     -16.794       0   
168246            0.000935      0.1650     -17.211       1   
168247            0.906000      0.1060     -16.431       0   

                                                   s_name  d_popularity  \
31                                        bab s uvula who           NaN   
1680     piano sonata no 2 op 35 i grave doppio movimento           0.0   
3921                                     sunless saturday          31.0   
3922                                         read my lips          35.0   
3923                                          rain forest          32.0   
...                                                   ...           ...   
168243                                            8 4 9 1           0.0   
168244                                               40 3           0.0   
168245                                               47 2           0.0   
168246  haydn string quartet no 52 in e flat major op ...           0.0   
168247         symphony no 1 op 9 iv con moto passacaglia           0.0   

             t_release_date  c_speechiness  c_tempo  c_valence s_zip Code  \
31      1995-01-18 00:00:00         0.0542   96.215     0.9720      74023   
1680    1928-08-19 00:00:00         0.0365   99.270     0.0775       7040   
3921    1985-03-10 00:00:00         0.1360  144.071     0.3960      65774   
3922    1985-11-24 00:00:00         0.0743  118.005     0.8980      92586   
3923    1985-02-27 00:00:00         0.0340  127.295     0.0634        918   
...                     ...            ...      ...        ...        ...   
168243  1935-01-25 00:00:00         0.9360   

In [59]:
spot['s_artists'] = spot['s_artists'].str.split(',').str[0].str.replace(r'[\\[\\]\']', '', regex=True)

C:\Users\Usuario\AppData\Local\Temp\ipykernel_5920\3149658578.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  spot['s_artists'] = spot['s_artists'].str.split(',').str[0].str.replace(r'[\\[\\]\']', '', regex=True)


In [60]:
spot['s_artists'] = spot['s_artists'].str.lower().str.replace(r'[^a-zA-Z0-9\s]', '', regex=True).str.strip()

C:\Users\Usuario\AppData\Local\Temp\ipykernel_5920\3056235362.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  spot['s_artists'] = spot['s_artists'].str.lower().str.replace(r'[^a-zA-Z0-9\s]', '', regex=True).str.strip()


In [64]:
spot['s_artists'].head()

31                              green day
1680    frederic chopin vladimir horowitz
3921                             fishbone
3922                       michael franks
3923                paul speer david lanz
Name: s_artists, dtype: object

In [65]:
top10 = spot['s_genero'].value_counts().nlargest(9).index.tolist()

In [66]:
spot.loc[~spot['s_genero'].isin(top10), 's_genero'] = 'other'


In [67]:
spot['s_genero'].value_counts()

s_genero
47 0     32980
20 0     30270
50 0     24401
12 0     22103
16 0     21910
21 0     21733
30 0      3232
49 0      3193
48 0      3117
other     1055
Name: count, dtype: int64

In [70]:
columns_to_add = ['zip', 'lat', 'lng', 'city', 'state_name']

In [71]:
zips_subset = zips[columns_to_add]

In [72]:
spot['s_zip Code'] = pd.to_numeric(spot['s_zip Code'], errors='coerce')
zips_subset['zip'] = pd.to_numeric(zips_subset['zip'], errors='coerce')

C:\Users\Usuario\AppData\Local\Temp\ipykernel_5920\2855800129.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  spot['s_zip Code'] = pd.to_numeric(spot['s_zip Code'], errors='coerce')
C:\Users\Usuario\AppData\Local\Temp\ipykernel_5920\2855800129.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  zips_subset['zip'] = pd.to_numeric(zips_subset['zip'], errors='coerce')


In [73]:
spot_merged = pd.merge(spot, zips_subset, left_on='s_zip Code', right_on='zip', how='left')


In [74]:
spot_merged.head()

acousticness                          s_artists  c_danceability  \
0      0.000186                          green day           0.494   
1      0.990000  frederic chopin vladimir horowitz           0.272   
2      0.000069                           fishbone           0.451   
3      0.068200                     michael franks           0.861   
4      0.936000              paul speer david lanz           0.605   

   c_duration_ms  c_energy                    s_id  c_instrumentalness  \
0         128200     0.932  2uarrcx71zl7jspun2kupz             0.00004   
1         473227     0.281  6wt8pknbsqnfqdoxgexvku             0.86700   
2         258440     0.970  4dtbf3w1v4nlzko6epwvue             0.10300   
3         222107     0.726  68lfky4tmwgkqfotnnez1j             0.00013   
4         258000     0.229  6fzfxmevhsdqkd8wjfp83n             0.86000   

   c_liveness  c_loudness  d_mode  ...  c_tempo  c_valence s_zip Code  \
0      0.0248      -4.606       0  ...   96.215     0.9720      74023   
1      0.0870     -16.228       0  ...   99.270     0.0775       7040   
2      0.1360      -4.622       1  ...  144.071     0.3960      65774   
3      0.0922     -12.759       1  ...  118.005     0.8980      92586   
4      0.1000     -16.300       1  ...  127.295     0.0634        918   

   s_genero  s_genero_normalizado    zip       lat        lng          city  \
0     other               2 3 3 1  74023  35.95802  -96.74985    Cus173hing   
1     other                         7040  40.73060  -74.26960   Map19lewood   
2      47 0                  47 0  65774  37.92914  -93.49222  We801aubleau   
3      20 0                  20 0  92586  33.70901 -117.19889    Menife875e   
4      50 0                  50 0    918  18.42138  -66.06588   San706 Juan   

       state_name  
0     Ok176lahoma  
1   New Jer303sey  
2     Misso551uri  
3   Cali972fornia  
4  Puerto Ri109co  

[5 rows x 24 columns]

In [75]:
spot_merged['lat'] = pd.to_numeric(spot_merged['lat'], errors='coerce')

In [76]:
spot_merged['lng'] = pd.to_numeric(spot_merged['lng'], errors='coerce')

In [77]:
nan_in_lat = spot_merged['lat'].isnull().sum()

In [78]:
nan_in_lng = spot_merged['lng'].isnull().sum()

In [79]:
nan_in_lat

np.int64(0)

In [80]:
nan_in_lng

np.int64(0)

In [81]:
spot_merged['city'] = spot_merged['city'].str.replace(r'\d', '', regex=True)


In [82]:
spot_merged['state_name'] = spot_merged['state_name'].str.replace(r'\d', '', regex=True)

In [83]:
spot_merged['state'] = spot_merged['city'].fillna('') + ', ' + spot_merged['state_name'].fillna('')

In [84]:
def remove_accents(input_str):
    nfkd_form = unicodedata.normalize('NFD', input_str)
    return u"".join([c for c in nfkd_form if not unicodedata.combining(c)])

In [85]:
spot_merged['state'] = spot_merged['state'].apply(remove_accents).str.lower()

In [86]:
invalid_dates_count = pd.to_datetime(spot_merged['t_release_date'], errors='coerce').isnull().sum()

In [87]:
spot_merged['t_release_date'] = pd.to_datetime(spot_merged['t_release_date'], errors='coerce')


In [88]:
spot_merged.dropna(subset=['t_release_date'], inplace=True)


In [89]:
spot_merged['t_release_date'].dtype

dtype('<M8[ns]')

In [90]:
genre_counts_series = spot_merged['s_genero'].value_counts()

In [91]:
genre_counts_df = genre_counts_series.reset_index()

In [92]:
genre_counts_df.columns = ['genero', 'conteo_registros']

In [93]:
genre_counts_df

genero  conteo_registros
0   47 0             32980
1   20 0             30270
2   50 0             24401
3   12 0             22103
4   16 0             21910
5   21 0             21733
6   30 0              3232
7   49 0              3193
8   48 0              3117
9  other              1055

In [94]:
spot_merged['duration_minutos'] = spot_merged['c_duration_ms'] / 60000

In [95]:
top_10_songs = spot_merged.sort_values(by='d_popularity', ascending=False).head(10)


In [96]:
top_10_songs[['s_name', 's_artists', 'd_popularity']]

s_name  \
82522                                  blinding lights   
82520                        rockstar feat roddy ricch   
82529  death bed coffee for your head feat beabadoobee   
82521                                       the scotts   
82425                                       supalonely   
82526                                     toosie slide   
82421                                          the box   
82433                  ily i love you baby feat emilee   
82533                  stuck with u with justin bieber   
82531                                            gooba   

                              s_artists  d_popularity  
82522                        the weeknd         100.0  
82520                dababy roddy ricch          99.0  
82529                 powfu beabadoobee          97.0  
82521  the scotts travis scott kid cudi          96.0  
82425               benee gus dapperton          95.0  
82526                             drake          95.0  
82421                       roddy ricch          95.0  
82433                  surf mesa emilee          95.0  
82533       ariana grande justin bieber          94.0  
82531                           6ix9ine          94.0

In [97]:
avg_milliseconds = spot_merged['c_duration_ms'].mean()

In [98]:
avg_minutes = spot_merged['duration_minutos'].mean()

In [99]:
avg_milliseconds

np.float64(231407.62518750687)

In [100]:
avg_minutes

np.float64(3.8567937531251144)

In [101]:
energy_by_genre_df = spot_merged.groupby('s_genero')['c_energy'].agg(['mean', 'count'])


In [102]:
energy_by_genre_df.columns = ['promedio_energia', 'conteo_registros']

In [103]:
energy_by_genre_df

promedio_energia  conteo_registros
s_genero                                    
12 0              0.493059             22103
16 0              0.488317             21910
20 0              0.488607             30270
21 0              0.489741             21733
30 0              0.488098              3232
47 0              0.491348             32980
48 0              0.493020              3117
49 0              0.489755              3193
50 0              0.492040             24401
other             0.480612              1055

In [104]:
min_loudness_row = spot_merged.loc[spot_merged['c_loudness'].idxmin()]

In [105]:
max_loudness_row = spot_merged.loc[spot_merged['c_loudness'].idxmax()]

In [106]:
min_loudness_row[['s_name', 's_artists', 'c_loudness']]

s_name          hava nagilah
s_artists     connie francis
c_loudness             -60.0
Name: 37055, dtype: object

In [107]:
max_loudness_row[['s_name', 's_artists', 'c_loudness']]

s_name                sold
s_artists     apocolothoth
c_loudness           3.855
Name: 102628, dtype: object

In [108]:
import numpy as np

continuous_vars = spot_merged.select_dtypes(include=np.number)
percentiles = [0.1, 0.2, 0.3, 0.4, 0.5, 0.6, 0.7, 0.8, 0.9, 1.0]
deciles = continuous_vars.describe(percentiles=percentiles)

In [109]:
deciles

acousticness  c_danceability  c_duration_ms       c_energy  \
count  163994.000000   163994.000000   1.639940e+05  163994.000000   
mean        0.490049        0.539222   2.314076e+05       0.490425   
std         0.376889        0.175448   1.210218e+05       0.267571   
min         0.000000        0.000000   5.108000e+03       0.000000   
10%         0.008710        0.299000   1.376130e+05       0.131000   
20%         0.052000        0.385000   1.625600e+05       0.222000   
30%         0.144000        0.448000   1.792670e+05       0.306000   
40%         0.290000        0.501000   1.936000e+05       0.395000   
50%         0.484000        0.549000   2.088000e+05       0.483000   
60%         0.681000        0.596000   2.275070e+05       0.571000   
70%         0.828000        0.643000   2.498000e+05       0.665000   
80%         0.932000        0.696000   2.797600e+05       0.760000   
90%         0.983000        0.761000   3.353870e+05       0.868000   
100%        0.996000        0.988000   5.403500e+06       1.000000   
max         0.996000        0.988000   5.403500e+06       1.000000   

       c_instrumentalness     c_liveness     c_loudness         d_mode  \
count       163994.000000  163994.000000  163994.000000  163994.000000   
mean             0.160891       0.206420     -11.327964       0.707941   
std              0.308620       0.176553       5.668497       0.454711   
min              0.000000       0.000000     -60.000000       0.000000   
10%              0.000000       0.072900     -18.873000       0.000000   
20%              0.000000       0.091100     -15.573400       0.000000   
30%              0.000002       0.105000     -13.459000       1.000000   
40%              0.000020       0.116000     -11.862000       1.000000   
50%              0.000193       0.135000     -10.420000       1.000000   
60%              0.001950       0.167000      -9.070000       1.000000   
70%              0.022800       0.222000      -7.730000       1.000000   
80%              0.276000       0.305000      -6.411000       1.000000   
90%              0.830000       0.407000      -4.988000       1.000000   
100%             1.000000       1.000000       3.855000       1.000000   
max              1.000000       1.000000       3.855000       1.000000   

        d_popularity  c_speechiness        c_tempo      c_valence  \
count  160605.000000  163994.000000  163994.000000  163994.000000   
mean       31.900128       0.094188     117.006147       0.532093   
std        21.642731       0.149761      30.739444       0.262279   
min         0.000000       0.000000       0.000000       0.000000   
10%         0.000000       0.030100      79.613300       0.159000   
20%         6.000000       0.033400      89.258000       0.270000   
30%        20.000000       0.036600      97.383900       0.367000   
40%        28.000000       0.040400     106.034000       0.460000   
50%        34.000000       0.045100     114.834000       0.544000   
60%        40.000000       0.052200     122.709800       0.626000   
70%        45.000000       0.064800     130.752000       0.708000   
80%        52.000000       0.093300     141.185000       0.793000   
90%        60.000000       0.194000     161.999000       0.887000   
100%      100.000000       0.969000     244.091000       1.000000   
max       100.000000       0.969000     244.091000       1.000000   

          s_zip Code            zip            lat            lng  \
count  163994.000000  163994.000000  163994.000000  163994.000000   
mean    49669.589015   49669.589015      38.848949     -91.209828   
std     27566.379204   27566.379204       5.338104      16.195823   
min       601.000000     601.000000      17.966120    -331.846760   
10%     12491.000000   12491.000000      32.092020    -115.968500   
20%     21613.000000   21613.000000      34.318620     -99.743820   
30%     30663.000000   30663.000000      36.432809     -95.612730   
40%     40946.000000   40946.000000      38.1

**Signo zodiacal del profesor: Aries
**Signo zodiacal del ayudante: Libra